File: testers/avro/java-spring/generated-and-checked-in/com/example/events/TreeNodePublisher.java
package com.example.events;

import java.util.concurrent.CompletableFuture;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.kafka.support.SendResult;
import org.springframework.stereotype.Service;

/** Type-safe event publisher for tree-node topic */
@Service
public record TreeNodePublisher(KafkaTemplate<String, TreeNode> kafkaTemplate, String topic) {
  public TreeNodePublisher(KafkaTemplate<String, TreeNode> kafkaTemplate) {
    this(kafkaTemplate, "tree-node");
  }

  public TreeNodePublisher withKafkaTemplate(KafkaTemplate<String, TreeNode> kafkaTemplate) {
    return new TreeNodePublisher(kafkaTemplate, topic);
  }

  public TreeNodePublisher withTopic(String topic) {
    return new TreeNodePublisher(kafkaTemplate, topic);
  }

  /** Publish a TreeNode event */
  public CompletableFuture<SendResult<String, TreeNode>> publish(String key, TreeNode event) {
    return kafkaTemplate.send(topic, key, event);
  }
}

File: testers/avro/scala-cats/generated-and-checked-in/com/example/events/TreeNodePublisher.scala
package com.example.events

import cats.effect.IO
import fs2.kafka.KafkaProducer
import fs2.kafka.ProducerRecord
import fs2.kafka.ProducerRecords
import fs2.kafka.ProducerResult
import scala.annotation.unused

@unused
/** Type-safe event publisher for tree-node topic */
case class TreeNodePublisher(
  kafkaTemplate: KafkaProducer[IO, String, TreeNode],
  topic: String = "tree-node"
) {
  /** Publish a TreeNode event */
  def publish(
    key: String,
    event: TreeNode
  ): IO[ProducerResult[String, TreeNode]] = {
    return kafkaTemplate.produce(ProducerRecords.one(ProducerRecord(topic, key, event))).flatten
  }
}